Migrate to using meson, deprecating autoconf

gonzoleeman opened the pull request:
Autoconf is brittle, so migrate to using the meson build system.

There should be no functional change. Autoconf support will go away soon.

gonzoleeman committed:
meson builds now working

This allows autotools to still work, but
meson build system is working too, in parallel.

gonzoleeman committed:
meson: Add ability to disable static library build

Also, ensure at least one type of library is built,
and use our libisns.ver library versioning file, when
building the shared lib, unless user has diabled that.

gonzoleeman committed:
Add a decprecation warning to configure script.

This prints a deprecation message when running the script.

gonzoleeman committed:
meson: update README with meson info

Updated a few other small things in the README, but mostly
add info about meson, and mention that autoconf (still explained)
is deprecated.

gonzoleeman committed:
meson: convert some args to 'features'

No real functional change, but features seem better than
true/false options?

gonzoleeman committed:
meson: fix error building shared lib with version

eli-schwartz commented on include/libisns/meson.build:
```diff
@@ -0,0 +1,27 @@
+  install_dir: get_option('includedir') / 'libisns')
+
+install_headers(static_hdr_files,
+  install_dir: get_option('includedir') / 'libisns')
```
```suggestion
  subdir: 'libisns')
```

This will automatically install within the includedir.

gonzoleeman replied:
turns out I didn't need subdir, either, since the files have "libisns/" prefixes.

eli-schwartz replied:
I think you do, unless you call the installation from `include/meson.build` so that the actual paths given to the function are stored as e.g. `'libisns/attrs.h'` instead of `'attrs.h'`, *and* use Meson 0.63.0 and `preserve_path: true`.

> Disable stripping child-directories from header files when installing.
> 
> This is equivalent to GNU Automake's nobase option.

eli-schwartz commented on meson.build:
```diff
@@ -0,0 +1,266 @@
+  )
+
+# get feature options
+with_security = get_option('security').enabled()
+with_slp = get_option('slp').enabled()
```
You're casting these to a boolean, true if enabled, false if auto (commonly called "check" in autotools land) / disabled.

The main advantage of a feature option, though, is that you can pass the option object itself as e.g. `security_dep = dependency('libcrypto', required: get_option('security'))`

In the enabled case, which you currently handle, this is equivalent: the dependency is mandatory when enabled.

In the disabled case, which you also handle, this is equivalent: the dependency is not even looked up. The UX is a bit different: when Meson *knows* there's a disabled feature, it prints "Dependency **libcrypto** skipped: feature **security** disabled".

In the auto case, you treat it like disabled, but Meson would look it up, try to use it if it's installed, but continue if it's not installed. autotools equivalent: `--with-security=check`.

In all cases, you can determine whether the security feature is ultimately merged into the build definitions with `if security_dep.found()`, or even:

```
conf.set('WITH_SECURITY', security_dep.found())
```

gonzoleeman replied:
cool. Fixed.

eli-schwartz commented on meson.build:
```diff
@@ -0,0 +1,266 @@
+  lib_deps += cc.find_library('slp')
+endif
+
+if enable_static
+  libisns_lib = static_library('isns',
+    lib_srcs,
+    include_directories: incl_dirs,
+    dependencies: lib_deps,
+    install: true)
+endif
+
+if enable_shared
+  if enable_shared_version
+    shared_link_args = '-Wl,--version-script=@0@/libisns.vers'.format(meson.current_source_dir())
+  else
+    shared_link_args = []
+  endif
+  libisns_lib = shared_library('isns',
+    lib_srcs,
+    include_directories: incl_dirs,
+    dependencies: lib_deps,
+    link_args: shared_link_args,
+    version: '0',
+    install: true)
+endif
```
I think you can just use `libisns_lib = library('isns', ...)`

It accepts the kwargs of both static and shared libraries, and can build one or both, depending on the value of `-Ddefault_library={static|shared|both}`. It will just use whichever arguments are relevant to whichever type it builds.

This is built into Meson, the same way --enable-shared and --enable-static is built into libtool.

gonzoleeman replied:
That seems much better -- except for the --version-script part. Is there something I can pass that will work all the time, i.e. meson figures out if it's okay or not? I looked for some 'version_script' argument to library() but didn't see one.

eli-schwartz replied:
It's just the same argument you'd pass to shared_library -- i.e. link_args. Those will be ignored if you never link, and static libraries don't link.

It's a goal of mine to eventually add a unified kwarg that replaces vs_module_defs (windows specific) with something that accepts any of:
- libtool flat text file of symbol names
- gnu ld version script
- Apple ld version script
- Windows defs file

And chooses (or converts) the best one.

However, that doesn't exist now.

(I think I'll be in a good position to implement it once some currently pending reworks of the build targets bit of the Meson interpreter go in. So: soon.)

gonzoleeman replied:
fixed

eli-schwartz commented on meson.build:
```diff
@@ -0,0 +1,266 @@
+# man pages
+#
+
+foreach m: man8_pages
+  install_man(m, install_dir: man_dir / 'man8')
+endforeach
+foreach m: man5_pages
+  install_man(m, install_dir: man_dir / 'man5')
+endforeach
```
install_man will cleverly figure out the correct subdir to install each man page, based on the extension (in this case, "8"). You can even add a list of man pages for several different sections as the arguments to install_man -- it will install each argument to the right place for that specific man page.

```suggestion
install_man(man8_pages, man5_pages)
```

gonzoleeman replied:
fixed

eli-schwartz commented on meson.build:
```diff
@@ -0,0 +1,266 @@
+#
+
+foreach c: config_files
+  install_data(c,
```
`install_data(config_files, install_dir: ...)`

gonzoleeman replied:
done

eli-schwartz commented on meson.build:
```diff
@@ -0,0 +1,266 @@
+endif
+
+#
+# create our runtime directory (empty)
+#
+
+install_subdir('isns',
+  install_dir: '/var/lib',
+  install_mode: 'r--------')
```
I would leave a comment here:

```
# this can be replaced with install_emptydir once the minimum requirement of Meson is bumped to >=0.60.0
```

Note that the use of install_subdir() is called out in the docs as "possible to use in this manner due to a clever loophole, but doing so is deprecated".

Of course, deprecated functionality is perfectly reasonable to use if you're still supporting older systems!

...

I'm not sure if the install_subdir hack supports install modes, I *think* that was one of the reasons that made me create install_emptydir instead.

gonzoleeman replied:
fixed

eli-schwartz commented on meson_options.txt:
```diff
@@ -0,0 +1,15 @@
+option('security', type: 'feature', value: 'enabled')
+option('slp', type: 'feature', value: 'enabled')
+
+# disable shared library creation and enable static library
```
If you use default_library instead, then you can just add `'default_library=static'` to the project default options.

gonzoleeman committed:
meson: several updates based on review

Several changes mostly suggested by Eli Schwartz based
on review from github:

- Require at least version 0.54 of meson
- use the "feature" interface correctly for 'slp' and 'security'
- no need to use 'foreach' for several types of installation
  calls, since the calls take a list
- man page installs figure out the subdirectory by themselves
- now using "--default-library={shared,static,both}" instead
  of our "enable_shared" and "enabled_static", letting meson
  figure it out

gonzoleeman committed:
meson: small option usage cleanup

No need to check options up front. Instead, chcck
them when needed.

gonzoleeman committed:
meson: update README

Almost completely meson-related, but also found one sentence
that was never finished, so I guessed what is supposed to
go there.

gonzoleeman committed:
Fix two compiler warnings in slp.c

They were both unused parameters. As long as I
was there, I removed spaces at the end of lines, as
well, since they drive me nuts.

gonzoleeman commented:
@eli-schwartz : I believe I addressed each of your suggestions/concerns, but please feel free to check, as I've pushed the updates.

Part of the message from one commit added:

>     Several changes mostly suggested by Eli Schwartz based
>     on review from github:
>     
>     - Require at least version 0.54 of meson
>     - use the "feature" interface correctly for 'slp' and 'security'
>     - no need to use 'foreach' for several types of installation
>       calls, since the calls take a list
>     - man page installs figure out the subdirectory by themselves
>     - now using "--default-library={shared,static,both}" instead
>       of our "enable_shared" and "enabled_static", letting meson
>       figure it out
> 

Also, I added a comment about installing an empty directory, and
I updated the README. Lastly, I fixed a compiler warning for some
unused variables in slp.c, just because the warning annoyed me.

I decided to just use the "--default-library=" option for controlling the
library type(s) created.

Thanks for all of your time and help!

gonzoleeman committed:
build: only specify version in one place

Whether using deprecated autotools or meson, use
the version number from the configuration file,
so we don't have to keep it in two places.

gonzoleeman committed:
meson: just specify subdir for header-file install.

Another fix suggested by Eli Schwartz (thanks!).

gonzoleeman commented:
> On Sep 5, 2022, at 4:52 PM, Eli Schwartz ***@***.***> wrote:
> 
> 
> @eli-schwartz commented on this pull request.
> 
> In include/libisns/meson.build <https://github.com/open-iscsi/open-isns/pull/39#discussion_r963156216>:
> 
> > +  'source.h',
> +  'types.h',
> +  'util.h'])
> +
> +template_hdr_name = 'paths.h'
> +template_hdr_file = files('paths.h.in')
> +
> +configure_file(
> +  input: template_hdr_file,
> +  output: template_hdr_name,
> +  configuration: inc_conf,
> +  install: true,
> +  install_dir: get_option('includedir') / 'libisns')
> +
> +install_headers(static_hdr_files,
> +  install_dir: get_option('includedir') / 'libisns')
> I think you do, unless you call the installation from include/meson.build so that the actual paths given to the function are stored as e.g. 'libisns/attrs.h' instead of 'attrs.h', and use Meson 0.63.0 and preserve_path: true.
> 
> Disable stripping child-directories from header files when installing.
> 
> This is equivalent to GNU Automake's nobase option.
> 
> 

So despite my comment, this is what I’m using now:

>	install_headers(static_header_files, subdir: ‘libisns’)

And it works correctly. When I didn’t do it correctly the build would fail, since the source code includes <libisns/FILE.h>.

gonzoleeman commented:
Any other objections? I'd like to merge this, since it doesn't break the current build system. Then I'd like to tag a new version.

going once ...

gonzoleeman commented:
Going going gone. Merging this request.
